File: scrapers/smc.py
import json
import os
from datetime import datetime, timedelta, timezone
from pathlib import Path

from playwright.sync_api import sync_playwright

# =========================
# CONFIG
# =========================

SOT_DIR = Path(
    os.getenv("LTL_MS_DIR", r"\\ant\dept-eu\BHX2\Public\AF-CST\LTL_MS")
) / "source_of_truth_crawler.csv"

USER_DATA_DIR = os.path.join(
    os.path.expanduser("~"), "AppData", "Local", "UserData"
)

SMC_URL = "https://smc-eu-dub.dub.proxy.amazon.com/orders/list/tab/1"
API_URL = "https://smc-eu-dub.dub.proxy.amazon.com/shipper/order/search"

HEADLESS = False
PAGE_SIZE = 200

# Rolling date window for the API filter (days relative to "now").
DATE_WINDOW_DAYS_BACK = 14
DATE_WINDOW_DAYS_FORWARD = 14

# Hard cap so a misconfigured filter can't pull the whole archive.
MAX_PAGES = 100

# Where to dump the captured live payload for debugging / discovering the
# date field name the API expects. Stored inside the workspace so it's easy
# to inspect.
DEBUG_DIR = Path(__file__).resolve().parent.parent / "debug"

# =========================
# HELPERS
# =========================

def _read_shipper_ids() -> list[str]:
    text = SOT_DIR.read_text()
    lines = text.strip().split("\n")
    columns = [c.strip().lower() for c in lines[0].split(",")]
    if "shipperid" not in columns:
        raise ValueError("source_of_truth_crawler.csv missing 'shipperid' column")
    idx = columns.index("shipperid")
    ids = []
    for line in lines[1:]:
        row = [c.strip() for c in line.split(",")]
        if len(row) > idx and row[idx]:
            ids.append(row[idx])
    return ids[:1000]


def _iso_window() -> tuple[str, str]:
    """Return (start, end) ISO-8601 timestamps for the rolling date window."""
    now = datetime.now(timezone.utc)
    start = now - timedelta(days=DATE_WINDOW_DAYS_BACK)
    end = now + timedelta(days=DATE_WINDOW_DAYS_FORWARD)
    fmt = "%Y-%m-%dT%H:%M:%S.000Z"
    return start.strftime(fmt), end.strftime(fmt)


def _build_payload(shipper_ids: list[str], page: int, template: dict | None = None) -> dict:
    """
    Build the search payload. If `template` (the live UI payload) is provided,
    use it as the base so we mirror sort order and any new filters
    automatically. Then override our explicit knobs.
    """
    base: dict = json.loads(json.dumps(template)) if template else {
        "sortCriteria": [{"field": "ORDER_CREATION_DATE", "sortDirection": "DESC"}],
        "andCriteria": {},
        "orCriteria": {},
        "pageCriteria": {"page": page, "size": PAGE_SIZE, "totalRecords": 0},
    }

    base.setdefault("sortCriteria", [{"field": "ORDER_CREATION_DATE", "sortDirection": "DESC"}])
    base.setdefault("orCriteria", {})
    and_c = base.setdefault("andCriteria", {})

    # Filters we always control.
    and_c["orderSources"] = ["SMC", "R4S", "EDI", "AFAPI"]
    and_c["freightTypes"] = ["LESS_THAN_TRUCKLOAD", "TRUCKLOAD", "INTERMODAL"]
    and_c["orderExecutionStatuses"] = [
        "IN_DRAFT", "NOT_PLANNED", "PENDING_CARRIER_ACCEPTANCE",
        "CARRIER_TENDER_ACCEPTED", "DRIVER_DISPATCHED", "LATE_TO_ARRIVE",
        "ARRIVED", "LATE_TO_DEPART", "DEPARTED",
        "PENDING_DELIVERY_CONFIRMATION", "DELIVERY_CONFIRMED",
        "PENDING_PAYMENT", "PAID", "CANCELLED", "REJECTED",
    ]
    # All countries — empty list means "no channel filter".
    and_c["shipperBusinessChannels"] = []
    and_c["shipperIds"] = shipper_ids
    # Don't restrict to the "ready for scheduling" bucket: orders that already
    # have a VRID assigned graduate out of that tab and would otherwise be
    # missed on subsequent syncs (causing stale VRIDs to linger in cst_runs).
    and_c.pop("readyForScheduling", None)

    # Date window: filter on origin stop date. Field/shape mirrors what the
    # SMC UI sends: andCriteria.originDateRange = {start, end} (ISO Z),
    # with a companion null `originDateRangeLabel`.
    start, end = _iso_window()
    and_c["originDateRangeLabel"] = None
    and_c["originDateRange"] = {"start": start, "end": end}

    base["pageCriteria"] = {"page": page, "size": PAGE_SIZE, "totalRecords": 0}
    return base


# =========================
# PUBLIC API
# =========================

def run_smc_export() -> list[dict] | None:
    """
    Fetches all orders from the SMC API using the persistent browser session
    (Amazon SSO auth) and returns them as a list of order dicts.
    Handles pagination automatically.
    """
    print(f"\n🚀 SMC API export started @ {datetime.now():%Y-%m-%d %H:%M}")

    shipper_ids = _read_shipper_ids()
    print(f"ℹ️ Loaded {len(shipper_ids)} shipper IDs")

    all_orders = []
    csrf_token = {"value": None}
    # Headers captured from the live SMC UI request to /shipper/order/search.
    # Used as the source of truth for replaying API calls — keeps us in sync
    # with whatever auth/CSRF scheme the UI is currently using.
    captured_headers: dict[str, str] = {}
    # Body captured from the live SMC UI request. Used as a payload template
    # so we mirror the date-range field name and any new filters automatically.
    captured_payload: dict = {}

    with sync_playwright() as p:
        browser = p.chromium.launch_persistent_context(
            user_data_dir=USER_DATA_DIR,
            headless=HEADLESS,
            # The Amazon corp proxy chain isn't trusted by Playwright's own
            # HTTP client. The browser itself trusts it via the OS store, so
            # navigation works, but `page.request.post(...)` fails with
            # "self-signed certificate in certificate chain" without this.
            ignore_https_errors=True,
        )
        page = browser.new_page()

        # Capture headers AND body from any request that the page makes to the
        # order-search endpoint. These are the canonical auth/CSRF headers and
        # the payload shape the UI is currently using.
        def on_request(request):
            if csrf_token["value"] is None:
                token = (
                    request.headers.get("x-csrf-token")
                    or request.headers.get("anti-csrftoken-a2z")
                )
                if token:
                    csrf_token["value"] = token
            if request.url.startswith(API_URL):
                if not captured_headers:
                    captured_headers.update(request.headers)
                if not captured_payload:
                    try:
                        body = request.post_data
                        if body:
                            captured_payload.update(json.loads(body))
                    except Exception:
                        pass

        page.on("request", on_request)
        page.goto(SMC_URL)
        page.wait_for_load_state("networkidle")

        # Fallback 1: wait actively for a request that carries the header,
        # in case the listing tab was served from cache and the listener
        # didn't fire during goto().
        if not csrf_token["value"]:
            try:
                req = page.wait_for_event(
                    "request",
                    predicate=lambda r: bool(
                        r.headers.get("x-csrf-token")
                        or r.headers.get("anti-csrftoken-a2z")
                    ),
                    timeout=15_000,
                )
                csrf_token["value"] = (
                    req.headers.get("x-csrf-token")
                    or req.headers.get("anti-csrftoken-a2z")
                )
            except Exception:
                pass

        # Fallback 2: pull it from the page's meta tags or cookies.
        if not csrf_token["value"]:
            try:
                csrf_token["value"] = page.evaluate(
                    """
                    () => {
                        const m = document.querySelector(
                            'meta[name="csrf-token"], meta[name="x-csrf-token"], meta[name="anti-csrftoken-a2z"]'
                        );
                        if (m) return m.getAttribute('content');
                        const cookie = document.cookie
                            .split(';')
                            .map(s => s.trim())
                            .find(s => /csrf|xsrf/i.test(s));
                        return cookie ? cookie.split('=').slice(1).join('=') : null;
                    }
                    """
                )
            except Exception:
                pass

        if not csrf_token["value"]:
            print("⚠️ Could not capture CSRF token — requests may be rejected")
        else:
            print(f"🔑 CSRF token captured ({len(csrf_token['value'])} chars)")

        # Trigger a real search if the page hasn't fired one yet, so we can
        # capture the exact request headers the UI uses (auth scheme can change
        # between subdomains and we want to mirror it precisely).
        if not captured_headers:
            try:
                page.wait_for_event(
                    "request",
                    predicate=lambda r: r.url.startswith(API_URL),
                    timeout=20_000,
                )
            except Exception:
                pass

        if captured_headers:
            print(f"📡 Captured {len(captured_headers)} headers from live SMC request")
        else:
            print("⚠️ No live SMC request observed — falling back to minimal headers")

        if captured_payload:
            and_keys = list((captured_payload.get("andCriteria") or {}).keys())
            print(f"🧬 Captured live payload — andCriteria keys: {and_keys}")
            try:
                DEBUG_DIR.mkdir(parents=True, exist_ok=True)
                dump_path = DEBUG_DIR / "smc_last_payload.json"
                dump_path.write_text(json.dumps(captured_payload, indent=2))
                print(f"   💾 Saved live payload to {dump_path}")
            except Exception as e:
                print(f"   ⚠️ Could not save debug payload: {e}")
        else:
            print("⚠️ No live SMC payload captured — using built-in template")

        win_start, win_end = _iso_window()
        print(
            f"📅 Date window (originDateRange): {win_start} → {win_end} "
            f"(±{DATE_WINDOW_DAYS_BACK}/{DATE_WINDOW_DAYS_FORWARD}d)"
        )

        current_page = 1

        while True:
            payload = _build_payload(shipper_ids, current_page, captured_payload or None)

            data = _post_via_browser(page, API_URL, payload, captured_headers, csrf_token["value"])
            if data is None:
                break

            orders = data.get("orders", [])
            all_orders.extend(orders)

            pr = data["pageResult"]
            total = pr["totalRecords"]
            print(f"  Page {current_page}: {len(orders)} orders (total: {total})")

            if current_page * PAGE_SIZE >= total:
                break
            if current_page >= MAX_PAGES:
                print(
                    f"⚠️ Hit MAX_PAGES={MAX_PAGES} cap at total={total}. "
                    f"Tighten the date window or filters if this looks wrong."
                )
                break
            current_page += 1

        browser.close()

    print(f"✅ Fetched {len(all_orders)} orders total")
    return all_orders if all_orders else None


def _post_via_browser(page, url: str, payload: dict, base_headers: dict, csrf: str | None):
    """
    POST `payload` to `url` from inside the browser using fetch(), so cookies,
    CSRF headers, and any session bootstrap come along the same as for the
    real UI. Returns the parsed JSON, or None on error.
    """
    # Build the headers we send: start from what the UI sends, then layer on
    # JSON content-type and (optionally) the captured CSRF token.
    headers: dict[str, str] = {}
    drop = {
        # Pseudo-headers and ones the browser will set itself.
        ":method", ":path", ":scheme", ":authority",
        "host", "content-length", "cookie", "connection",
    }
    for k, v in (base_headers or {}).items():
        if k.lower() in drop:
            continue
        headers[k] = v
    headers.setdefault("content-type", "application/json")
    headers.setdefault("accept", "application/json, text/plain, */*")
    if csrf:
        headers.setdefault("x-csrf-token", csrf)
        headers.setdefault("anti-csrftoken-a2z", csrf)

    try:
        result = page.evaluate(
            """
            async ({url, headers, body}) => {
                const resp = await fetch(url, {
                    method: "POST",
                    credentials: "include",
                    headers,
                    body: JSON.stringify(body),
                });
                const text = await resp.text();
                return { status: resp.status, body: text };
            }
            """,
            {"url": url, "headers": headers, "body": payload},
        )
    except Exception as e:
        print(f"❌ Browser fetch threw: {e}")
        return None

    status = result.get("status")
    if status != 200:
        snippet = (result.get("body") or "")[:300]
        print(f"❌ API error: HTTP {status} — {snippet}")
        return None

    try:
        return json.loads(result["body"])
    except Exception as e:
        print(f"❌ Could not parse JSON response: {e}")
        return None

File: extension/content/smc.js
/**
 * SMC data source — content-script side.
 *
 * Runs on the SMC page, so it fetches /shipper/order/search same-origin with the
 * page's own session cookies (credentials:"include"). This is the read source of
 * truth for the load list (SharePoint only holds the manual-sourcing work we do
 * on top).
 *
 * It is the SMC BRIDGE for the standalone UI (ui/app.html): the background
 * (smcClient.js) finds/opens an SMC tab and sends `smc:*` messages, answered at
 * the bottom of this file. window.__ltlSmc is kept for console diagnostics.
 */
(function () {
  "use strict";

  const SEARCH_URL = "https://smc-eu-dub.dub.proxy.amazon.com/shipper/order/search";
  const PAGE_SIZE = 25;
  const MAX_PAGES = 100;
  const DAYS_BACK = 7;
  const DAYS_FORWARD = 7;

  // Raw orders from the most recent fetch, kept for diagnostics.
  let __lastRawOrders = [];

  function dlog(...a) {
    if (window.__ltlDebug && window.__ltlDebug.status && window.__ltlDebug.status.call)
      console.debug("[LTL smc]", ...a);
    else console.debug("[LTL smc]", ...a);
  }

  // ── date window: default one week back / one week forward, ".000Z" ──────────
  function defaultWindow() {
    const now = Date.now();
    const fmt = (ms) => new Date(ms).toISOString().replace(/\.\d+Z$/, ".000Z");
    return { start: fmt(now - DAYS_BACK * 86400_000), end: fmt(now + DAYS_FORWARD * 86400_000) };
  }

  // Normalize a caller-supplied window into ISO-Z bounds.
  //  - start/end as "YYYY-MM-DD"  → start = 00:00:00Z, end = 23:59:59Z of that day
  //  - start/end as a full ISO datetime (contains "T") → used as-is
  // Missing/invalid start or end falls back to the default bound.
  function resolveWindow(win) {
    const def = defaultWindow();
    if (!win || (!win.start && !win.end)) return def;
    const norm = (v, kind, fallback) => {
      if (!v) return fallback;
      const s = String(v);
      // Full datetime provided → use verbatim (parse to normalize the format).
      if (s.includes("T")) {
        const d = new Date(s);
        return Number.isNaN(d.getTime()) ? fallback : d.toISOString().replace(/\.\d+Z$/, ".000Z");
      }
      // Date-only → start/end of that UTC day.
      const time = kind === "start" ? "T00:00:00.000Z" : "T23:59:59.000Z";
      const d = new Date(`${s.slice(0, 10)}${time}`);
      return Number.isNaN(d.getTime()) ? fallback : d.toISOString().replace(/\.\d+Z$/, ".000Z");
    };
    return { start: norm(win.start, "start", def.start), end: norm(win.end, "end", def.end) };
  }

  // ── query (team-scoped) ─────────────────────────────────────────────────────
  // Historical LTL query, used when no team query is supplied.
  const DEFAULT_QUERY = {
    orderSources: ["SMC", "EDI", "R4S", "AFAPI", "AFDIG"],
    freightTypes: ["LESS_THAN_TRUCKLOAD"],
    orderExecutionStatuses: [
      "IN_DRAFT", "PENDING_CARRIER_ACCEPTANCE", "CARRIER_TENDER_ACCEPTED",
      "DRIVER_DISPATCHED", "LATE_TO_ARRIVE", "ARRIVED", "LATE_TO_DEPART",
      "DEPARTED", "PENDING_DELIVERY_CONFIRMATION", "DELIVERY_CONFIRMED",
      "PENDING_PAYMENT", "PAID", "REJECTED", "NOT_PLANNED",
    ],
    shipperBusinessChannels: ["DE", "GB"],
    readyForScheduling: true,
  };
  // SMC caps shipperIds per request (CST_viewer used 1000).
  const MAX_SHIPPER_IDS = 1000;

  /**
   * Build the /shipper/order/search payload.
   * @param {number} page
   * @param {{start,end}} win
   * @param {object} [opts]
   * @param {object} [opts.query]      andCriteria overrides (Config.TEAMS[x].smcQuery)
   * @param {string[]} [opts.shipperIds]  restrict to these shipper IDs (CST SoT)
   */
  function buildPayload(page, win, opts = {}) {
    const { start, end } = resolveWindow(win);
    const q = { ...DEFAULT_QUERY, ...(opts.query || {}) };
    const andCriteria = {
      orderSources: q.orderSources,
      freightTypes: q.freightTypes,
      orderExecutionStatuses: q.orderExecutionStatuses,
      shipperBusinessChannels: q.shipperBusinessChannels || [],
      originDateRangeLabel: null,
      originDateRange: { start, end },
    };
    // null/undefined = don't send the readyForScheduling filter at all.
    if (q.readyForScheduling != null) andCriteria.readyForScheduling = q.readyForScheduling;
    if (opts.shipperIds && opts.shipperIds.length) {
      andCriteria.shipperIds = opts.shipperIds.slice(0, MAX_SHIPPER_IDS);
    }
    return {
      sortCriteria: [{ field: "ORDER_CREATION_DATE", sortDirection: "DESC" }],
      andCriteria,
      orCriteria: {},
      pageCriteria: { page, size: opts.pageSize || PAGE_SIZE },
    };
  }

  // ── best-effort CSRF token from the live page ───────────────────────────────
  function captureCsrf() {
    const meta = document.querySelector(
      'meta[name="csrf-token"], meta[name="x-csrf-token"], meta[name="anti-csrftoken-a2z"]'
    );
    if (meta) return meta.getAttribute("content");
    const cookie = document.cookie
      .split(";")
      .map((s) => s.trim())
      .find((s) => /csrf|xsrf/i.test(s));
    return cookie ? cookie.split("=").slice(1).join("=") : null;
  }

  async function postSearch(payload, csrf) {
    const headers = {
      "content-type": "application/json",
      accept: "application/json, text/plain, */*",
    };
    if (csrf) {
      headers["x-csrf-token"] = csrf;
      headers["anti-csrftoken-a2z"] = csrf;
    }
    const res = await fetch(SEARCH_URL, {
      method: "POST",
      credentials: "include",
      headers,
      body: JSON.stringify(payload),
    });
    dlog(`POST search page → HTTP ${res.status}`);
    if (res.status !== 200) {
      const snippet = (await res.text().catch(() => "")).slice(0, 300);
      throw new Error(`SMC API HTTP ${res.status} — ${snippet}`);
    }
    return res.json();
  }

  // ── shipper price (revenue) ─────────────────────────────────────────────────
  // Port of CST_viewer scripts/smc_update.py _extract_revenue(): prefer the
  // LINE_HAUL line, else the first priced line.
  function pricingLines(order) {
    return ((order.shipperPricing || {}).pricing) || [];
  }
  function preferredLine(order) {
    const lines = pricingLines(order);
    if (!lines.length) return null;
    return lines.find((p) => p && p.type === "LINE_HAUL") || lines[0];
  }
  function extractRevenue(order) {
    const line = preferredLine(order);
    const v = line && line.price ? line.price.value : null;
    const n = v == null ? NaN : parseFloat(v);
    return Number.isNaN(n) ? null : n;
  }
  // Currency varies by channel (GBP/EUR), so read it rather than assuming.
  function extractRevenueCurrency(order) {
    const line = preferredLine(order);
    const p = (line && line.price) || {};
    const c = p.currencyCode || p.currency || p.currencyUnit || null;
    return c ? String(c).trim().toUpperCase() : null;
  }

  // ── order -> one row per VRID ───────────────────────────────────────────────
  function orderToRows(order) {
    const stops = order.stops || [];
    const stop1 = stops[0] || {};
    const stop2 = stops.length > 1 ? stops[stops.length - 1] : {};
    const shipper = order.shipperDetails || {};
    const carrierDetails = order.carrierDetails || {};
    const origAddr = stop1.address || {};
    const destAddr = stop2.address || {};

    const base = {
      orderid: String(order.orderIdentifier?.id ?? ""),
      shipperid: String(shipper.shipperId ?? "").trim(),
      shippername: shipper.shipperName,
      shipper_ref: order.shipperReferenceId,
      // sourcing signals (SMC has no RLB1/DUMMY/AZNG — "needs sourcing" = no carrier yet)
      carrier_offer_count: order.carrierOfferCount ?? 0,
      has_carrier: !!(carrierDetails && (carrierDetails.carrierId || carrierDetails.carrierName)),
      order_status: order.orderStatus,
      scheduling_status: order.schedulingStatus,
      orig_country: origAddr.countryCode,
      dest_country: destAddr.countryCode,
      origin: stop1.stopName,
      dest: stop2.stopName,
      lane: `${stop1.stopName || ""} → ${stop2.stopName || ""}`,
      origin_code: stop1.stopLocationCode,
      dest_code: stop2.stopLocationCode,
      orig_node: stop1.stopLocationCode,
      dest_node: stop2.stopLocationCode,
      equipment_type: order.equipmentType,
      freight_type: order.freightType,
      vehicle_carrier: carrierDetails.carrierId,
      orig_planned_yard_checkin_time: stop1.startTime,
      dest_planned_yard_checkin_time: stop2.startTime,
      status: order.orderStatus,
      execution_status: order.executionStatus,
      vehicle_execution_status: order.vrExecutionStatus || order.executionStatus,
      isa: stop2.appointmentId,
      // Shipper price — what the margin calculator rates the carrier quote
      // against. Ports CST_viewer smc_update.py _extract_revenue().
      revenue: extractRevenue(order),
      revenue_currency: extractRevenueCurrency(order),
    };

    const vrids = (order.vehicleRunIds || []).filter(Boolean);
    const list = vrids.length ? vrids : [null];
    return list.map((vrid) => ({ ...base, vrid }));
  }

  /**
   * Fetch every order in the window (all pages), return flat rows (one per VRID).
   * `opts` is forwarded to buildPayload ({query, shipperIds}).
   */
  async function fetchRows(win, opts = {}) {
    const csrf = captureCsrf();
    dlog(csrf ? `csrf captured (len=${csrf.length})` : "no csrf token found on page");
    const w = resolveWindow(win);
    dlog(`window ${w.start} → ${w.end}`);
    const rows = [];
    const size = opts.pageSize || PAGE_SIZE;
    __lastRawOrders = [];
    let page = 1;
    while (true) {
      const data = await postSearch(buildPayload(page, w, opts), csrf);
      const orders = data.orders || [];
      __lastRawOrders.push(...orders);
      for (const o of orders) rows.push(...orderToRows(o));
      const total = data.pageResult?.totalRecords ?? orders.length;
      dlog(`page ${page}: ${orders.length} orders (total ${total}), rows so far ${rows.length}`);
      if (page * size >= total || page >= MAX_PAGES || orders.length === 0) break;
      page += 1;
    }
    return rows;
  }

  /**
   * Keep only rows that "still need sourcing", per the team's sourcing knobs
   * (Config.TEAMS[x].sourcing):
   *   requireVrid          manual sourcing is per-VRID, so no VRID = not a candidate
   *   requireNoCarrier     no carrier assigned in SMC (RLB1/AZNG/DUMMY placeholders
   *                        live in FMC, not SMC)
   *   excludeFreightTypes  e.g. CST drops LESS_THAN_TRUCKLOAD (LTL team owns those)
   */
  const DEFAULT_SOURCING = { requireVrid: true, requireNoCarrier: true, excludeFreightTypes: [] };

  function filterSourcing(rows, sourcing = DEFAULT_SOURCING) {
    const s = { ...DEFAULT_SOURCING, ...(sourcing || {}) };
    const excluded = new Set((s.excludeFreightTypes || []).map((v) => String(v).toUpperCase()));
    return rows.filter((r) => {
      if (s.requireVrid && String(r.vrid ?? "").trim() === "") return false;
      if (s.requireNoCarrier && r.has_carrier) return false;
      if (excluded.size && excluded.has(String(r.freight_type ?? "").toUpperCase())) return false;
      return true;
    });
  }

  /**
   * Fetch within `win` ({start,end}) + filter to orders still needing sourcing.
   * @param {{start,end}} win
   * @param {object} [opts]
   * @param {object} [opts.query]       andCriteria overrides for the team
   * @param {string[]} [opts.shipperIds] shipper allow-list (team SoT)
   * @param {object} [opts.sourcing]    post-filter knobs for the team
   * @param {object} [opts.shipperMap]  { shipperid: {shipper_group,...} } → tags rows
   */
  async function fetchSourcingRows(win, opts = {}) {
    const all = await fetchRows(win, { query: opts.query, shipperIds: opts.shipperIds });
    const filtered = filterSourcing(all, opts.sourcing);
    if (opts.shipperMap) {
      for (const r of filtered) {
        const s = opts.shipperMap[String(r.shipperid ?? "").trim()];
        r.shipper_group = s ? s.shipper_group : "";
      }
    }
    dlog(`fetched ${all.length} rows, ${filtered.length} still need sourcing`);
    return filtered;
  }

  /**
   * Look up specific orders/VRIDs regardless of sourcing state (e.g. a run that
   * already has a real carrier and so isn't on the list).
   *
   * SMC's search API has no by-ID filter we know of, so: (1) check the raw
   * orders from the most recent fetch (already in memory, free); (2) if any ID
   * is still missing, re-pull `win` (caller passes a wide window) with the
   * team's query at a large page size and match again.
   * @returns {{ rows, found: string[], missing: string[], source: "cache"|"smc" }}
   */
  async function lookupByIds(ids, win, opts = {}) {
    const wanted = new Set(ids.map((v) => String(v).trim()).filter(Boolean));
    if (!wanted.size) return { rows: [], found: [], missing: [], source: "cache" };
    const matches = (rows) =>
      rows.filter((r) => wanted.has(String(r.orderid ?? "").trim()) || wanted.has(String(r.vrid ?? "").trim()));
    const foundIds = (rows) => {
      const s = new Set();
      for (const r of rows) {
        if (wanted.has(String(r.orderid ?? "").trim())) s.add(String(r.orderid).trim());
        if (wanted.has(String(r.vrid ?? "").trim())) s.add(String(r.vrid).trim());
      }
      return s;
    };

    let rows = matches(__lastRawOrders.flatMap(orderToRows));
    let source = "cache";
    if (foundIds(rows).size < wanted.size) {
      dlog(`lookup: ${wanted.size - foundIds(rows).size} id(s) not in cache — re-pulling window`);
      const all = await fetchRows(win, { query: opts.query, shipperIds: opts.shipperIds, pageSize: 200 });
      rows = matches(all);
      source = "smc";
    }
    if (opts.shipperMap) {
      for (const r of rows) {
        const s = opts.shipperMap[String(r.shipperid ?? "").trim()];
        r.shipper_group = s ? s.shipper_group : "";
      }
    }
    const found = [...foundIds(rows)];
    const missing = [...wanted].filter((id) => !found.includes(id));
    dlog(`lookup: ${rows.length} row(s), found ${found.length}, missing ${missing.length} (${source})`);
    return { rows, found, missing, source };
  }

  // ── logged-in user (requester) from SMC config ──────────────────────────────
  // Same-origin fetch of /configuration/constants; the `requester` field is the
  // signed-in alias (e.g. "mayowas"). Cached after first success.
  let __requester = null;
  async function getRequester() {
    if (__requester) return __requester;
    try {
      const res = await fetch(
        "https://smc-eu-dub.dub.proxy.amazon.com/configuration/constants",
        { credentials: "include", headers: { accept: "application/json" } }
      );
      if (res.status !== 200) {
        dlog(`getRequester: HTTP ${res.status}`);
        return null;
      }
      const data = await res.json();
      __requester = data && data.requester ? String(data.requester).trim() : null;
      dlog(`getRequester: ${__requester || "(none)"}`);
      return __requester;
    } catch (e) {
      dlog(`getRequester failed: ${e.message}`);
      return null;
    }
  }

  // ── bridge: answer the background's smc:* messages ──────────────────────────
  // The UI is a standalone extension page (ui/app.html), so it can't fetch SMC
  // itself. The background finds/opens an SMC tab and sends these; we run the
  // same-origin fetch here and reply { bridge:true, ok, ... }.
  const fail = (err) => ({
    bridge: true,
    ok: false,
    status: (err && err.status) || 0,
    expired: !!(err && err.expired),
    error: String(err && err.message ? err.message : err),
  });

  async function ping() {
    const res = await fetch(
      "https://smc-eu-dub.dub.proxy.amazon.com/configuration/constants",
      { credentials: "include", headers: { accept: "application/json" } }
    );
    const ctype = (res.headers.get("content-type") || "").toLowerCase();
    if (res.redirected && !res.url.startsWith(location.origin)) {
      const e = new Error(`SMC session expired — redirected to ${res.url}`);
      e.status = 401; e.expired = true; throw e;
    }
    if (res.status === 401 || res.status === 403 || (res.status === 200 && !ctype.includes("json"))) {
      const e = new Error(`SMC session expired — HTTP ${res.status} (${ctype || "no content-type"})`);
      e.status = res.status === 200 ? 401 : res.status; e.expired = true; throw e;
    }
    if (res.status !== 200) {
      const e = new Error(`SMC ping HTTP ${res.status}`);
      e.status = res.status; throw e;
    }
    return res.status;
  }

  browser.runtime.onMessage.addListener((msg) => {
    if (!msg || typeof msg.action !== "string" || !msg.action.startsWith("smc:")) return;
    dlog(`bridge ← ${msg.action}`);
    if (msg.action === "smc:ping") {
      return ping().then((status) => ({ bridge: true, ok: true, status })).catch(fail);
    }
    if (msg.action === "smc:requester") {
      return getRequester().then((requester) => ({ bridge: true, ok: true, requester })).catch(fail);
    }
    if (msg.action === "smc:lookup") {
      return lookupByIds(msg.ids || [], msg.win || {}, msg.opts || {})
        .then((r) => ({ bridge: true, ok: true, ...r }))
        .catch((err) => {
          if (/HTTP (401|403)|redirected|Unexpected token/i.test(String(err && err.message))) {
            err.expired = true;
            err.status = err.status || 401;
          }
          return fail(err);
        });
    }
    if (msg.action === "smc:sourcingRows") {
      return fetchSourcingRows(msg.win || {}, msg.opts || {})
        .then((rows) => ({ bridge: true, ok: true, rows }))
        .catch((err) => {
          // A 200 that isn't JSON / a redirect means the session lapsed.
          if (/HTTP (401|403)|redirected|Unexpected token/i.test(String(err && err.message))) {
            err.expired = true;
            err.status = err.status || 401;
          }
          return fail(err);
        });
    }
    return fail(new Error(`Unknown smc action: ${msg.action}`));
  });
  browser.runtime.sendMessage({ action: "smc:bridge-ready", href: location.href }).catch(() => {});

  window.__ltlSmc = {
    fetchRows,
    fetchSourcingRows,
    filterSourcing,
    getRequester,
    rawOrders: () => __lastRawOrders,
  };
})();
